src/control/server/storage/scm.go

```diff
@@ -0,0 +1,499 @@
+	op := fmt.Sprintf(msgMount, devPath, mntPoint, mntType, fmt.Sprint(flags), opts)
+
+	e.log.Debugf(op)
+	e.appendHistory(op)
+
+	if flags == 0 {
+		flags = uintptr(syscall.MS_NOATIME | syscall.MS_SILENT)
+		flags |= syscall.MS_NODEV | syscall.MS_NOEXEC | syscall.MS_NOSUID
+	}
+
+	if err := syscall.Mount(devPath, mntPoint, mntType, flags, opts); err != nil {
+		return errPermsAnnotate(os.NewSyscallError("mount", err))
+	}
+	return nil
+}
+
+func (e *scmExtImpl) unmount(path string) error {
+	e.log.Debugf(msgUnmount, path)
+	e.appendHistory(fmt.Sprintf(msgUnmount, path))
+
+	// ignore NOENT errors, treat as success
+	if err := syscall.Unmount(
+		path, syscall.MNT_DETACH); err != nil && !os.IsNotExist(err) {
+
+		// when mntpoint exists but is unmounted, get EINVAL
+		e, ok := err.(syscall.Errno)
+		if ok && e == syscall.EINVAL {
+			return nil
+		}
+
+		return errPermsAnnotate(os.NewSyscallError("umount", err))
+	}
+	return nil
+}
+
+func (e *scmExtImpl) remove(path string) error {
+	e.log.Debugf(msgRemove, path)
+	e.appendHistory(fmt.Sprintf(msgRemove, path))
+
+	// ignore NOENT errors, treat as success
+	if err := os.RemoveAll(path); err != nil && !os.IsNotExist(err) {
+		return errPermsAnnotate(errors.WithMessage(err, "remove"))
+	}
+	return nil
+}
+
+func defaultScmExt(log logging.Logger) *scmExtImpl {
+	return &scmExtImpl{
+		log:     log,
+		history: []string{},
+	}
+}
+
+// ScmProvider gives access to underlying storage interface implementation
+// for accessing SCM devices (API) in addition to storage of device
+// details.
+//
+// IpmCtl provides necessary methods to interact with Storage Class
+// Memory modules through libipmctl via ipmctl bindings.
+type ScmProvider struct {
+	log         logging.Logger
+	ext         scmExt
+	ipmctl      ipmctl.IpmCtl // ipmctl NVM API interface
+	prep        PrepScm
+	modules     types.ScmModules
+	initialized bool
+	formatted   bool
+}
+
+// TODO: implement remaining methods for ScmProvider
+// func (s *ScmProvider) Update(req interface{}) interface{} {return nil}
+// func (s *ScmProvider) BurnIn(req interface{}) (fioPath string, cmds []string, env string, err error) {
+// return
+// }
+
+// Setup implementation for ScmProvider providing initial device discovery
+func (s *ScmProvider) Setup() error {
+	return s.Discover()
+}
+
+// Teardown implementation for ScmProvider
+func (s *ScmProvider) Teardown() error {
+	s.initialized = false
+	return nil
+}
+
+// Prep configures pmem kernel devices for SCM
+func (s *ScmProvider) Prep(state types.ScmState) (needsReboot bool, PmemDevs []PmemDev, err error) {
+	return s.prep.Prep(state)
+}
+
+// PrepReset resets configuration of SCM
+func (s *ScmProvider) PrepReset(state types.ScmState) (needsReboot bool, err error) {
+	return s.prep.PrepReset(state)
+}
+
+// Discover method implementation for ScmProvider
+func (s *ScmProvider) Discover() error {
+	if s.initialized {
+		return nil
+	}
+
+	mms, err := s.ipmctl.Discover()
```

Comment:
please make sure the changes from https://github.com/daos-stack/daos/pull/1079 get carried over into this PR if it lands before

Reply:
done